contracts/RandomBeaconHistory.cdc

```diff
@@ -22,10 +22,58 @@ access(all) contract RandomBeaconHistory {
     access(contract) var lowestHeight: UInt64?
     /// Sequence of random sources recorded by the Heartbeat, stored as an array over a mapping to reduce storage
     access(contract) let randomSourceHistory: [[UInt8]]
+    /// Start index of the first gap in the `randomSourceHistory` array where random sources were not recorded, because
+    /// of a heartbeat failure.
+    /// There may be non contiguous gaps in the history, `gapStartIndex` is the start index of the lowest-height
+    /// gap.
+    /// If no gap exists, `gapStartIndex` is equal to the `randomSourceHistory` array length.
+    access(contract) var gapStartIndex: UInt64
```

Comment:
As mentioned in the Discord [discussion](https://discord.com/channels/613813861610684416/1210247669683986442/1222263812359716944), the contract can't be updated by adding a new field, so an alternative solution needs to be figured out for this field. 
Note that this field is added because we limit backfilling to a maximum number of entries per transaction (to limit the total computation in the transaction/block). If we allow backfilling to go through all missing entries without a limit, the logic won't need to store this field. 

This means we have 2 options:
 - update the logic to allow unlimited backfilling and remove this field. Note that backfilling is only needed in the very rare case where the heartbeat transaction fails, and this won't always result in a large computation block.
 - keep the current logic and decide where to save the new field. The opinion of someone more familiar with Cadence is appreciated in this case. Should another contract be added to store this field?

Reply:
WRT dealing with the limitation on adding fields, I think we could deal with this a couple of ways.

1. Add a contract `ServiceAccountReservedFields` which contains mappings of `{String: AnyStruct}` and `@{String: AnyResource}` and add values to those mappings across any service account contracts that may need it.
2. Define a new struct `GapStartIndex` with a single optional variable `value: UInt64` and save that in the service account at some storage path (`/storage/randomBeaconHistoryGapStartIndex`) from which the contract could access and update where needed.
3. Define a new resource `Backfiller` and offload the entirety of the gap state tracking and filling task to that resource. We'd move the `gapStartIndex` field to that resource along with the backfilling logic, all of which are [legal updates](https://cadence-lang.org/docs/1.0/language/contract-updatability#structs-resources-and-interfaces). During every SoR commitment, we would then borrow a reference to the resource and call `backfill()` which would execute the backfill logic if its required at that moment.

IMO 3. is cleaner and is easier to reason about looking at the contract

Reply:
Opened #417 against this branch with the changes suggested in 3 for feedback